libs/s25main/figures/nofMiner.cpp

```diff
@@ -71,20 +72,100 @@ helpers::OptionalEnum<GoodType> nofMiner::ProduceWare()
 
 bool nofMiner::AreWaresAvailable() const
 {
+    // FindPointWithResource triggeres outofresource message
+    if(GetAddonSetting() == 3)
+        return true;
+
     return nofWorkman::AreWaresAvailable() && FindPointWithResource(GetRequiredResType()).isValid();
 }
 
-bool nofMiner::StartWorking()
+unsigned int nofMiner::GetAddonSetting() const 
 {
-    MapPoint resPt = FindPointWithResource(GetRequiredResType());
-    if(!resPt.isValid())
-        return false;
     const GlobalGameSettings& settings = world->GetGGS();
-    bool inexhaustibleRes = settings.isEnabled(AddonId::INEXHAUSTIBLE_MINES)
-                            || (workplace->GetBuildingType() == BuildingType::GraniteMine
-                                && settings.isEnabled(AddonId::INEXHAUSTIBLE_GRANITEMINES));
-    if(!inexhaustibleRes)
-        world->ReduceResource(resPt);
+
+    switch(workplace->GetBuildingType())
+    {
+        case BuildingType::GoldMine: return settings.getSelection(AddonId::MINES_GOLD);
+        case BuildingType::IronMine: return settings.getSelection(AddonId::MINES_IRON);
+        case BuildingType::CoalMine: return settings.getSelection(AddonId::MINES_COAL);
+        case BuildingType::GraniteMine: return settings.getSelection(AddonId::MINES_GRANITE);
+        default: return 0;
+    }
+}
+
+bool nofMiner::StartWorking()
+{
+    workplace->is_emptyCycle = false;
+
+    // needs to have at least one resource spot
+    // 0 = is exhaustible
+    // 1 = settlers IV like
+    // 2 = inexhaustible
+    // 3 = everywhere
+    unsigned int addonSettings = GetAddonSetting();
+
+    switch (addonSettings)
+    {
+        case 1: // settlers IV style
+        {
+            int sumResAmount = 0;
+            MapPoint useResPt;
+
+            std::vector<MapPoint> resPts = FindAllPointsWithResource(GetRequiredResType());
+
+            // iterate over all points
+            for each(MapPoint curPt in resPts)
+            {
+                // calculate the absolute amount of resource beneath
+                uint8_t resAmount = world->GetNode(curPt).resources.getAmount();
+                sumResAmount += resAmount;
+
+                // if there is one with more than 1 quantity, keep it (for reducing)
+                if(resAmount > 1 && !useResPt.isValid())
+                    useResPt = curPt;
+            }
+
+            // no resources left (mine was built on invalid spot)
+            if(sumResAmount == 0)
+                return false;
+
+            // 19 = amount of nodes a mine can reach
+            // 7  = maximum resource amount a node possibly has
+            if(RANDOM.Rand(RANDOM_CONTEXT(), 19 * 7) > sumResAmount)
```

Comment:
To many magic numbers for my liking. Maybe put the contents of the `FindAllPointsWithResource` here and first get all points in the given radius then iterate over them. Then you can use the size of the vector of all points here (I guess that is the 19?)
This also avoids calling `world->GetNode(curPt).resources` 2 times per node (one in `HasResource` one above)

And for the `7`: There should be an (enforced and named) constant for that. If not, the `Resource` class should get one.
And there is a shortcut for this line:
```suggestion
            if(RANDOM_RAND(19 * 7) > sumResAmount)
```
See https://github.com/Return-To-The-Roots/s25client/blob/fd16c75ecc85ef814885ce9b1fd41d1e701fa6cd/libs/s25main/random/Random.h#L99